### PR TITLE
p2p: Avoid multiple getheaders messages in flight to the same peer

### PR DESCRIPTION
**Pull request by sdaftuar**
Change `getheaders` messages so that we wait up to 2 minutes for a response to a prior `getheaders` message before issuing a new one.

Also change the handling of the `getheaders` message sent in response to a block INV, so that we no longer use the hashstop variable (including the hash stop will just mean that if our peer's headers chain is longer, then we won't learn
it, so there's no benefit to using hashstop).

Also, now respond to a `getheaders` during IBD with an empty headers message (rather than nothing) -- this better conforms to the intent of the new logic that it's better to not ignore a peer's `getheaders` message, even if you have nothing to give. This also avoids a lot of functional tests breaking.

This PR also reworks the headers processing logic to make it more readable.

### CODE REVIEW

**Comment by sdaftuar** [comment]
@ajtowns This should fix the issue you've seen in functional tests, where generating a bunch of blocks at once can result in O(n^2) headers downloads.

**Review by joani24 — APPROVED** [review]

**Comment by laanwj** [comment]
Concept ACK, great catch

File: src/net_processing.cpp [hunk 1]
```diff
@@ -3403,6 +3412,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (m_chainman.ActiveTip() == nullptr ||
                 (m_chainman.ActiveTip()->nChainWork < nMinimumChainWork && !pfrom.HasPermission(NetPermissionFlags::Download))) {
             LogPrint(BCLog::NET, "Ignoring getheaders from peer=%d because active chain has too little work\n", pfrom.GetId());
```
Comment (dergoegge) [inline src/net_processing.cpp:3414]: Maybe change this log message? because now we send something in response instead of ignoring

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2258,23 +2265,19 @@ bool PeerManagerImpl::CheckHeadersAreContinuous(const std::vector<CBlockHeader>&
...
- * pindexLast should be the last header we learned from a peer in their prior
- * headers message.
- *
- * This is used for headers sync with a peer; even if pindexLast is an ancestor
- * of a known chain (such as our tip) we don't yet know where the peer's chain
- * might fork from what we know, so we continue exactly from where the peer
- * left off.
- */
-void PeerManagerImpl::FetchMoreHeaders(CNode& pfrom, const CBlockIndex *pindexLast, const Peer& peer)
+bool PeerManagerImpl::MaybeSendGetHeaders(CNode& pfrom, const CBlockLocator& locator, Peer& peer)
```
Comment (dergoegge) [inline src/net_processing.cpp:2268]: Why did you change it to receive a `CBlockLocator`? I don't think we ever create locators that are not part of the main chain. Also (maybe a bit nitpicky) we dont need to compute the locator if we don't send the `getheaders` message, so doing that inside of `MaybeSendGetHeaders` makes more sense to me.

**Review by dergoegge — COMMENTED** [review]
Concept ACK

File: src/net_processing.cpp (hunk 2 above)
Comment (sdaftuar) [reply src/net_processing.cpp:2268]: The reason I introduced this interface change here is that this branch is a precursor to another change I will be proposing to the headers sync logic, in which we might send getheaders messages based on locators that are not calculated from a single CBlockIndex entry.

File: src/net_processing.cpp (hunk 1 above)
Comment (sdaftuar) [reply src/net_processing.cpp]: Fixed.

**Review by mzumsande — COMMENTED** [review]
Concept ACK

File: src/net_processing.cpp [hunk 3]
```diff
@@ -2443,15 +2443,9 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
...
-
-        // If we don't have the last header, then they'll have given us
-        // something new (if these headers are valid).
-        if (!m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash())) {
-            received_new_header = true;
-        }
-    }
+    // If we don't have the last header, then this peer will have given us
+    // something new (if these headers are valid).
+    received_new_header = WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash()) == nullptr);
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:2451]: We need to do the calculation here before invoking `ProcessNewBlockHeaders()` because that will cause headers to be added to the block index, but I'll get rid of the declaration at line 2423, as we can just declare and initialize here instead.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -4423,16 +4443,17 @@ void PeerManagerImpl::ConsiderEviction(CNode& pto, std::chrono::seconds time_in_
...
+                if (MaybeSendGetHeaders(pto,
+                            m_chainman.ActiveChain().GetLocator(state.m_chain_sync.m_work_header->pprev), *peer)) {
+                    LogPrint(BCLog::NET, "sending getheaders to outbound peer=%d to verify chain work (current best known block:%s, benchmark blockhash: %s)\n", pto.GetId(), state.pindexBestKnownBlock != nullptr ? state.pindexBestKnownBlock->GetBlockHash().ToString() : "<none>", state.m_chain_sync.m_work_header->GetBlockHash().ToString());
+                    state.m_chain_sync.m_sent_getheaders = true;
+                    // Bump the timeout to allow a response, which could clear the timeout
+                    // (if the response shows the peer has synced), reset the timeout (if
+                    // the peer syncs to the required work but not to our tip), or result
+                    // in disconnect (if we advance to the timeout and pindexBestKnownBlock
+                    // has not sufficiently progressed)
+                    state.m_chain_sync.m_timeout = time_in_seconds + HEADERS_RESPONSE_TIME;
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:4455]: Thanks for thinking about this.  I think you're right to raise this concern, and after re-reviewing this logic, I think the best course of action is to just have this call-site assume the `getheaders` message goes out (rather than branch on the return value) -- because if it doesn't, it means we've already sent a `getheaders`, and so the peer should have provided us with a sufficiently high work chain anyway. 

Does that reasoning sound right to you as well?

[the remark this answers is not in the document]
Comment (mzumsande) [reply src/net_processing.cpp]: > Does that reasoning sound right to you as well?

Yes, that's also the solution I had thought of.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -2206,55 +2424,35 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
-        }
+    // At this point, the headers connect to something in our block index.
+    if (!CheckHeadersAreContinuous(headers)) {
+        Misbehaving(peer, 20, "non-continuous headers sequence");
+        return;
     }
 
+    // If we don't have the last header, then this peer will have given us
+    // something new (if these headers are valid).
+    bool received_new_header = WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash()) == nullptr);
```
Comment (dergoegge) [inline src/net_processing.cpp:2454]:
```suggestion
    bool received_new_header{WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash()) == nullptr)};
```

File: src/net_processing.cpp [hunk 6]
```diff
@@ -2194,6 +2215,203 @@ void PeerManagerImpl::SendBlockTransactions(CNode& pfrom, Peer& peer, const CBlo
...
+        hashLastBlock = header.GetHash();
+    }
+    return true;
+}
+
+bool PeerManagerImpl::MaybeSendGetHeaders(CNode& pfrom, const CBlockLocator& locator, Peer& peer)
+{
+    const CNetMsgMaker msgMaker(pfrom.GetCommonVersion());
+
+    const auto current_time = GetTime<std::chrono::seconds>();
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp]: Thanks, I tried to make this switch in the last commit but it took me several tries to get it to compile, so please review and let me know if I'm using `NodeClock` correctly!

File: src/net_processing.cpp [hunk 7]
```diff
@@ -560,6 +565,22 @@ class PeerManagerImpl final : public PeerManager
                                const std::vector<CBlockHeader>& headers,
                                bool via_compact_block)
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
+    /** Various helpers for headers processing, invoked by ProcessHeadersMessage() */
+    /** Deal with state tracking and headers sync for peers that send the
+     * occasional non-connecting header (this can happen due to BIP 130 headers
+     * announcements for blocks interacting with the 2hr rule). */
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:571]: One of the most unusual validation rules is that we don't accept a block that is more than 2 hours in the future, see:
 https://github.com/bitcoin/bitcoin/blob/5bf65ec66e5986c9188e3f6234f1c5c0f8dc7f90/src/validation.cpp#L3469

File: src/net_processing.cpp [hunk 8]
```diff
@@ -2206,55 +2424,35 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
+    if (!headers_connect_blockindex) {
+        if (nCount <= MAX_BLOCKS_TO_ANNOUNCE) {
+            // If this looks like it could be a BIP 130 block announcement, use
+            // special logic for handling headers that don't connect, as this
+            // could be benign.
+            HandleFewUnconnectingHeaders(pfrom, peer, headers);
+        } else {
+            BlockValidationState state;
+            state.Invalid(BlockValidationResult::BLOCK_MISSING_PREV, "prev-blk-not-found");
+            MaybePunishNodeForBlock(pfrom.GetId(), state, via_compact_block, "invalid header received");
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp]: Done.

File: src/net_processing.cpp (hunk 5 above)
Comment (sdaftuar) [reply src/net_processing.cpp]: Took this change and also switched to this type of initialization for  `headers_connect_blockindex` as well.

File: src/net_processing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp]: Dropping the std::atomic seems to have made things work with both clang and g++, so I think this is probably right now (thanks to @dongcarl and @sipa for helping me sort through this!).

File: src/net_processing.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net_processing.cpp:574]: Ah ok, thanks for clarifying! Maybe mention the `MAX_FUTURE_BLOCK_TIME` constant here to make it obvious?

File: src/net_processing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net_processing.cpp]: Interesting, the `std::atomic` version did work for me locally (compiled with clang). What you have now looks good to me!

[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp]: @MarcoFalke I think I'll probably give up on trying to use `std::atomic` because all my efforts seem, but if you feel like writing the code and can share a diff that compiles I'd be happy to include it.

File: src/net_processing.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:574]: Done!

File: src/net_processing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp]: I presume the issue was g++-8 in https://cirrus-ci.com/task/5966736871653376?logs=ci#L2017

I tried my proposal (calling `NodeSeconds{}`) and it worked on top of your commit.

```diff
# git log -1 && git diff 
commit fa24d37213cede7136ed34d481091b84642c62a6 (HEAD)
Author: Suhas Daftuar <sdaftuar@gmail.com>
Date:   Tue Jun 28 10:53:02 2022 -0400

    Replace GetTime() with NodeClock in MaybeSendGetHeaders()
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index a75e3bd7f..6fc3009cb 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -358,7 +358,7 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
     /** Time of the last getheaders message to this peer */
-    std::atomic<NodeClock::time_point> m_last_getheaders_timestamp{};
+    std::atomic<NodeClock::time_point> m_last_getheaders_timestamp{NodeSeconds{}};
 
     Peer(NodeId id)
         : m_id{id}
```

An equivalent would be to call `NodeClock::time_point{}` instead.

**Comment by dergoegge** [comment]
Code review ACK b6f63bd991569d45b19c4f2d83b223e6e7cdcff3

File: src/net_processing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp]: I think the problem here is `std::atomic<T>` requires `T` to be trivially copyable, but unlike `duration`, `time_point` isn't implemented as trivially copyable (and none of `time_point`, `duration`, or the underlying `rep` are specified as being trivially copyable either). Since we already assume `duration` is trivially copyable, last time I messed with this I wrote a wrapper, something like:

```c++
template<typename TP>
class atomic_time_point {
private:
    using dur = typename TP::duration;
    std::atomic<dur> d;
public:
    atomic_time_point() noexcept : d{dur{0}} { }
    atomic_time_point(const atomic_time_point&) = delete;
    atomic_time_point(const TP& t) noexcept : d{t.time_since_epoch()} { }
    TP load() { return TP{d.load()}; }
    TP operator=(const TP& t) { d = t.time_since_epoch(); return t; }
};
using AtomicNodeTime = atomic_time_point<NodeClock::time_point>;
using AtomicNodeSeconds = atomic_time_point<NodeSeconds>;
```

which would let you write `AtomicNodeTime m_last_getheaders_timestamp{};` without relying on `time_point` implementation details.

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp]: > time_point isn't implemented as trivially copyable

Can you add references to your claim that they are not trivially copyable?

See also the cpp program which passes on gcc7/clang5:

```cpp
#include <chrono>

template <class T>
void CanUseInAtomic() {
    static_assert(std::is_trivially_copyable_v<T>);
    static_assert(std::is_copy_constructible_v<T>);
    static_assert(std::is_move_constructible_v<T>);
    static_assert(std::is_copy_assignable_v<T>);
    static_assert(std::is_move_assignable_v<T>);
}

int main() {
    CanUseInAtomic<bool>();
    CanUseInAtomic<std::chrono::seconds>();      // duration
    CanUseInAtomic<std::chrono::system_clock::time_point>();  // time_point
}
```



https://godbolt.org/z/3nKc43cPv

[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp]: Hmm, fair point. `static_assert(std::is_nothrow_default_constructible_v<T>);` catches it for clang 5, but not for gcc 7.1 for what that's worth.

But I don't think the pre-C++20 behaviour of default `std::atomic` initialization is ever what we want anyway? Per https://en.cppreference.com/w/cpp/atomic/atomic/atomic

> 1) The default constructor is trivial: no initialization takes place other than [zero initialization](https://en.cppreference.com/w/cpp/language/zero_initialization) of static and thread-local objects. [std::atomic_init](https://en.cppreference.com/w/cpp/atomic/atomic_init) may be used to complete initialization.

that will just leave things undefined, like `int x;` ?

(Anyway, no need for further changes in this PR, as far as I can see)

[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp]: I can't say I'm exactly following everything you guys are talking about, but @MarcoFalke thank you -- I updated the PR and the compilers all seem happy now. :). Will squash that last fixup commit now...

**Comment by ajtowns** [comment]
> @ajtowns This should fix the issue you've seen in functional tests, where generating a bunch of blocks at once can result in O(n^2) headers downloads.

Can confirm that it seems to fix the problem, and that the debug logs for p2p_blockheaders show much fewer "getheaders" requests.

**Comment by ajtowns** [comment]
ACK 99f4785cad94657dcf349d00fdd6f1d44cac9bb0 ; code review, check over new logic of when to send getheaders messages

File: test/functional/feature_minchainwork.py [hunk 9]
```diff
@@ -82,7 +82,7 @@ def run_test(self):
         msg.hashstop = 0
         peer.send_and_ping(msg)
         time.sleep(5)
-        assert "headers" not in peer.last_message
+        assert ("headers" not in peer.last_message or len(peer.last_message["headers"].headers) == 0)
```
Comment (sipa) [inline test/functional/feature_minchainwork.py:85]: Nit: braces are unnecessary, I think.

**Review by sipa — COMMENTED** [review]
utACK 99f4785cad94657dcf349d00fdd6f1d44cac9bb0

Nits are very non-blocking

**Comment by dergoegge** [comment]
Code review ACK  99f4785cad94657dcf349d00fdd6f1d44cac9bb0

**Review by w0xlt — APPROVED** [review]
tACK https://github.com/bitcoin/bitcoin/pull/25454/commits/99f4785cad94657dcf349d00fdd6f1d44cac9bb0 Good improvement in the code.

Tested on signet.

**Comment by mzumsande** [comment]
Code Review ACK 99f4785cad94657dcf349d00fdd6f1d44cac9bb0

**Review by maflcko — COMMENTED** [review]
left some questions

File: src/net_processing.cpp [hunk 10]
```diff
@@ -2208,36 +2255,24 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
+    bool headers_connect_blockindex{WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers[0].hashPrevBlock) != nullptr)};
+
+    if (!headers_connect_blockindex) {
+        if (nCount <= MAX_BLOCKS_TO_ANNOUNCE) {
+            // If this looks like it could be a BIP 130 block announcement, use
+            // special logic for handling headers that don't connect, as this
+            // could be benign.
+            HandleFewUnconnectingHeaders(pfrom, peer, headers);
+        } else {
+            Misbehaving(peer, 10, "invalid header received");
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp:2439]: It still gets a score of 20 via `HandleFewUnconnectingHeaders()` when `nCount <= MAX_BLOCKS_TO_ANNOUNCE`; this adds an unavoidable score of 10 for unconnecting headers when `nCount > MAX_BLOCKS_TO_ANNOUNCE` which previously was more or less ignored...

[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net_processing.cpp:2439]: Previously a non-connecting non-continuous header sequence of size MAX_UNCONNECTING_HEADERS would be submitted to validation through `ProcessNewBlockHeaders()`, since the `if (!m_chainman.m_blockman.LookupBlockIndex(headers[0].hashPrevBlock) && nCount <= MAX_BLOCKS_TO_ANNOUNCE)` block would be skipped.

That would return `BlockValidationResult::BLOCK_MISSING_PREV`:

https://github.com/bitcoin/bitcoin/blob/691a08718beff31d1b821b192609ea3bfdb24d41/src/validation.cpp#L3590-L3593

which would result in 10 misbehaviour points in `MaybePunishNodeForBlock`

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp:2439]: > this adds an unavoidable score of 10 for unconnecting headers when nCount > MAX_BLOCKS_TO_ANNOUNCE which previously was more or less ignored...

Can you explain a bit more why this was "more or less ignored"? 

Previously it was deterministically and consistently assigned a `Misbehaving(peer, 20, "non-continuous headers sequence");`.

Now it is deterministically and consistently assigned a `Misbehaving(peer, 10, "invalid header received");`.

> Previously a non-connecting non-continuous header sequence of size MAX_UNCONNECTING_HEADERS would be submitted to validation through ProcessNewBlockHeaders()

No, it wouldn't, as it would be rejected by the `Misbehaving(peer, 20, "non-continuous headers sequence");` check?


What am I missing?

[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net_processing.cpp:2439]: > What am I missing?

Nothing. I misread non-continuous as continuous. You're right that this would be rejected with `Misbehaving(peer, 20, "non-continuous headers sequence")`

File: src/net_processing.cpp [hunk 11]
```diff
@@ -358,7 +358,7 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
     /** Time of the last getheaders message to this peer */
-    std::atomic<std::chrono::seconds> m_last_getheaders_timestamp{0s};
+    std::atomic<NodeClock::time_point> m_last_getheaders_timestamp{NodeSeconds{}};
```
Comment (MarcoFalke) [inline src/net_processing.cpp:361]: Is there even a need to make this atomic? In a single threaded environment this is not needed. And if the same peer is served by several threads (for whatever reason) in the future, an atomic may prevent UB, but doesn't seem sufficient to prevent races logically.

File: src/net_processing.cpp [hunk 12]
```diff
@@ -2194,6 +2199,48 @@ void PeerManagerImpl::SendBlockTransactions(CNode& pfrom, Peer& peer, const CBlo
     m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::BLOCKTXN, resp));
 }
 
+/**
+ * Special handling for unconnecting headers that might be part of a block
+ * announcement.
+ *
+ * We'll send a getheaders message in response to try to connect the chain.
+ *
+ * The peer can send up to MAX_UNCONNECTING_HEADERS in a row that
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:2224]: Sure, I can update the comment in a followup (there's a sense in which what is written is true, in that we only are checking the first header in a message to see if it connects, but that's probably an overly convoluted way to think about things).

File: src/net_processing.cpp (hunk 11 above)
Comment (sdaftuar) [reply src/net_processing.cpp:361]: Hmm that is an interesting point.  My sense was that we don't want to rely on implicit single-threadedness to reason about data races, so it was safer to just throw this into an atomic and not worry about it.

I also think that having multiple threads servicing the same peer would not really work for a bunch of reasons (including the one you give, about logical errors in code like this).  I would imagine that multithreading network handling in the future would involve different threads servicing different peers.

I think the most likely form of UB from data races would be if we were to expose this value via RPC in the future.  So to make this future-proof, I'd prefer that we leave this as-is; it doesn't seem like the performance hit of having an atomic here (which is rarely accessed, as `getheaders` messages are not a frequent occurrence, and there are far more effective ways to CPU DoS a node than to trigger a getheaders) is significant enough to warrant making this harder to reason about in the future.

File: src/net_processing.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:2439]: Not sure what to say here other than that yes, there are minor behavior changes in this PR.  I don't think any are significant as the DoS points here are pretty arbitrary and when checking for different failures, it shouldn't matter too much which order we do things.  I would have been happy to update the commit message as you suggest but now that this is merged there's not much I can do!

It should be the case that any failure before that resulted in DoS points would still result in DoS points now, I think -- if that is not the case, then that would be an oversight to fix.  The motivation for making these kinds of minor changes is to make the headers processing easier to reason about, which is also in preparation for another change I'll be proposing to headers sync in the future (which I think will be easier to understand with the headers processing logic that is introduced in this PR).

File: src/net_processing.cpp [hunk 13]
```diff
@@ -2444,15 +2443,9 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
-
-        // If we don't have the last header, then they'll have given us
-        // something new (if these headers are valid).
-        if (!m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash())) {
-            received_new_header = true;
-        }
-    }
+    // If we don't have the last header, then this peer will have given us
+    // something new (if these headers are valid).
+    bool received_new_header{WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers.back().GetHash()) == nullptr)};
```
[the remark this answers is not in the document]
Comment (sdaftuar) [reply src/net_processing.cpp:2452]: I suppose!  My view is that the performance here is extremely minor compared to code readability and organizing the ideas in this function together, so I prefer it this way.

File: src/net_processing.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp:2439]: Yeah, it wasn't meant as a criticism of https://github.com/bitcoin/bitcoin/commit/7f2450871b3ea0b4d02d56bd2ca365fcc25cf90e more as a nit to keep in mind in the future. I think the code changes in this commit are perfectly fine. Generally I just prefer to split behaviour changes from refactoring changes, or at least explain the behaviour changes in the commit message briefly. Otherwise, what seems obvious to the author is easily missed by reviewers. (I think this discussion is supporting evidence enough).

File: src/net_processing.cpp (hunk 11 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:361]: I think if this was ever exposed on RPC, we'd also want to think which other fields to return atomically in the response, in which case a per-field atomic isn't enough to prevent logic races. So to me it seems slightly preferable to leave the design of thread safety to when it is needed. Otherwise it may be assumed in the future that this is already perfectly thread safe and can be used as is. Leaving this plain (without atomic) would at least have a thread sanitizer failure hint that the design needs to be re-thought if this was accessed by more than one thread in the future.

Comment (MarcoFalke) [reply src/net_processing.cpp:361]: (Other than that, agree that atomic doesn't affect performance)

File: src/net_processing.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp:2224]: Did that in https://github.com/bitcoin/bitcoin/pull/25555. Feel free to NACK or ACK

File: src/net_processing.cpp (hunk 11 above)
Comment (sdaftuar) [reply src/net_processing.cpp:361]: You know, I think you're right that this atomic is incorrect.  We need to hold a lock throughout `MaybeSendGetHeaders()` because this variable can be accessed in the scheduler thread or in the net processing thread, and if we don't hold a lock throughout that function we can get a data race.  Thanks for asking about this, I'll open a PR to fix.

Comment (sdaftuar) [reply src/net_processing.cpp:361]: Addressed in #25557

Comment (MarcoFalke) [reply src/net_processing.cpp:361]: Can you elaborate a bit on this? Unless I am missing something `MaybeSendGetHeaders`/`ConsiderEviction` is only called in-thread, not to be confused with `CheckForStaleTipAndEvictPeers`, which is called out-of-thread. What am I missing?

Comment (vasild) [reply src/net_processing.cpp:361]: What is "in-thread" and "out-of-thread"?

Comment (MarcoFalke) [reply src/net_processing.cpp:361]: "in-thread" means running in the thread "ThreadMessageHandler (b-msghand): Application level message handling (sending and receiving). Almost all net_processing and validation logic runs on this thread."

Comment (vasild) [reply src/net_processing.cpp:361]: Right, `m_last_getheaders_timestamp` is accessed from `PeerManagerImpl::MaybeSendGetHeaders()` and `PeerManagerImpl::ProcessMessage()` which are only called from a single thread `ThreadMessageHandler()`.

> because this variable can be accessed in the scheduler thread or in the net processing thread

@sdaftuar, I think it is not accessed from the scheduler thread?

I am in favor of simplicity, in this case meaning to not protect data that is accessed by a single thread. If in the future it is accessed by another thread, then an appropriate protection should be added. That may be an atomic or a mutex just for that variable or a mutex for multiple variables depending on the need. Otherwise we may end up protecting it needlessly now and later to have to rework the protection anyway (or worse - use the available, inappropriate protection, like @MarcoFalke mentioned above "...it may be assumed in the future that this is already perfectly thread safe...").

Comment (sdaftuar) [reply src/net_processing.cpp:361]: > Can you elaborate a bit on this? Unless I am missing something MaybeSendGetHeaders/ConsiderEviction is only called in-thread, not to be confused with CheckForStaleTipAndEvictPeers, which is called out-of-thread. What am I missing?

Ah, I guess I misremembered how `ConsiderEviction` is called (and thought it was run in the scheduler thread).  

I'll update my other PR and drop the mutex/atomic altogether.

Comment (ryanofsky) [reply src/net_processing.cpp:361]: > I am in favor of simplicity, in this case meaning to not protect data that is accessed by a single thread.

Possible idea for the future. In theory could have `SingleThread<T>` class similar to the `Synced<T>` class from #25390 that wraps an object and asserts in debug mode that is always accessed from the same thread. I think a solution like that would be overkill here, and I'm struggling to think of cases where it wouldn't be overkill, but it wouldn't be hard to implement and could be useful to help code document assumptions it is making.

File: src/net_processing.cpp [hunk 14]
```diff
@@ -2194,6 +2199,48 @@ void PeerManagerImpl::SendBlockTransactions(CNode& pfrom, Peer& peer, const CBlo
...
 }
 
+/**
+ * Special handling for unconnecting headers that might be part of a block
+ * announcement.
+ *
+ * We'll send a getheaders message in response to try to connect the chain.
+ *
+ * The peer can send up to MAX_UNCONNECTING_HEADERS in a row that
+ * don't connect before given DoS points.
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/net_processing.cpp:2225]: Done in https://github.com/bitcoin/bitcoin/pull/25555

File: test/functional/feature_minchainwork.py (hunk 9 above)
Comment (MarcoFalke) [reply test/functional/feature_minchainwork.py:85]: Done as part of 8efa73e7ce4ea0c1b7ad5c3947a5ecf9fb6361d3

File: src/net_processing.cpp (hunk 11 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:361]: Should be resolved by 8efa73e7ce4ea0c1b7ad5c3947a5ecf9fb6361d3

File: src/net_processing.cpp [hunk 15]
```diff
@@ -2206,55 +2424,33 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
...
-        // MAX_BLOCKS_TO_ANNOUNCE), use special logic for handling headers that
-        // don't connect:
-        // - Send a getheaders message in response to try to connect the chain.
-        // - The peer can send up to MAX_UNCONNECTING_HEADERS in a row that
-        //   don't connect before giving DoS points
-        // - Once a headers message is received that is valid and does connect,
-        //   nUnconnectingHeaders gets reset back to 0.
-        if (!m_chainman.m_blockman.LookupBlockIndex(headers[0].hashPrevBlock) && nCount <= MAX_BLOCKS_TO_ANNOUNCE) {
-            nodestate->nUnconnectingHeaders++;
-            m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::GETHEADERS, m_chainman.ActiveChain().GetLocator(m_chainman.m_best_header), uint256()));
```
Comment (Sjors) [inline src/net_processing.cpp:2225]: msgMaker is now unused in ProcessHeadersMessage. This is fixed in #25717

**Comment by hebasto** [comment]
> `msgMaker` is now unused in ProcessHeadersMessage.

Same for `PeerManagerImpl::HandleFewUnconnectingHeaders()` and `PeerManagerImpl::ConsiderEviction()`.
See https://github.com/bitcoin/bitcoin/pull/27368.
